# Los enteros de numpy


Veamos los resultados de las siguientes operaciones, de los enteros de numpy, sobre 'este' computador. Es probable que si ejecuta este cuaderno en un computador diferente los resultados difieran. Es por eso que bajo los resultados de las operaciones se copiaron los resultados obtenidos originalmente. 

In [58]:
import numpy as np
A=np.matrix([[2,2**10],[2**20,2**30]])
A

matrix([[         2,       1024],
        [   1048576, 1073741824]])

In [59]:
2*A

matrix([[          4,        2048],
        [    2097152, -2147483648]])

`matrix([[          4,        2048],
        [    2097152, -2147483648]])`

## ¿-2147483648? ¡Negativo!

In [60]:
4*A

matrix([[      8,    4096],
        [4194304,       0]])

`matrix([[      8,    4096],
        [4194304,       0]])
`
## ¿0? ¡Cero!

In [61]:
8*A

matrix([[     16,    8192],
        [8388608,       0]])

`matrix([[     16,    8192],
        [8388608,       0]])
`
## ¿0? Cero

In [62]:
(2**32)*A


matrix([[         8589934592,       4398046511104],
        [   4503599627370496, 4611686018427387904]], dtype=int64)

`matrix([[         8589934592,       4398046511104],
        [   4503599627370496, 4611686018427387904]], dtype=int64)
`
## ¡Está bien!
No puede multiplicar por 2 ni por 4, pero sí puede multiplicar por $2^{32}$ 

## !¿Qué está pasando?!

Resulta que los enteros que usa numpy no son los mismos enteros que usa Python. Los enteros de numpy van desde $-2^{31}$ hasta $2^{31}-1$. Lo malo es que cuando una operación sobrepasa estos límites no se genera ningún tipo de advertencia y puede llevar a serios problemas. ¿y por qué se permite esto? en aras de hacer más rápido los cálculos.

## ¿Y por qué cuando se multiplica por  $2^{32}$ sí funciona?
Porque para representar $2^{32}$ numpy tiene que usar un formato diferente a los enteros, entonces cambia de formato para poder representar estos números. 

## ¿Qué podemos hacer?
Por un lado, no podemos usar los flotantes porque aproximan. Por otro lado, los enteros no permiten números muy grandes. Para completar los fraccionarios no se imprimen bien. Entonces, es hora de sospechar que esta no es la mejor librería para nuestros objetivos. Numpy está pensado para hacer cálculos aproximados de grandes matrices a gran velocidad. Pero nosotros necesitamos valores exactos de pequeñas matrices. Para este objetivo la librería adecuada es sympy. Afortunadamente realiza algunas operaciones de manera similar. Sin embargo, hay algunas diferencias:
* En vez de usar `@` para la multiplicación matricial, utiliza `*`.
* En vez de usar `matrix` para para definir una matriz, utiliza `Matrix`.
* En vez de usar `Fraction` para para definir racional, utiliza `sp.Rational`. Pero este se puede omitir si la fracción se  realiza después de otra operación con algún objeto de sympy. Los detalles de cuando se puede omitir `Rational` son tema de la materia Programación Orientada a Objetos.
* En vez de intercambiar renglones con `A[[0,1],:] = A[[1,0],:]` utiliza `A.row_swap(0,1)`.

De ahora en adelante usaremos sympy en vez de numpy.

In [63]:
import sympy as sp

# _Recorderis_

## Definición.
Si se puede pasar de una matriz $A$ a una matriz $B$ por medio de operaciones elementales de dice que las matrices $A$ y $B$ son **equivalentes**.[Nakos, sec 1.2, pg 17]

## Teorema.
Una matriz puede ser equivalente a muchas matrices <u>escalón</u>. Pero sólo es equivalente a una matriz <u>escalón reducida</u>.

Los ejercicios que siguen, requieren haber visto las presentaciones de las secciones 3.1 y 3.2 del libro de Algebra Lineal de Nakos. 

# Ejercicio 1
De la sección 3.1, página 167, responda los siguientes ejercicios.
(Voy a responder el ejercicio con datos ficticios para ilustrar <u>la forma en que deben</u> colocar la respuesta.)

In [64]:
ejercicio_2_a_x=0
ejercicio_2_a_y=0
ejercicio_2_a_z=0
ejercicio_4_a='ACEG'
ejercicio_4_b='ABCDH'
ejercicio_4_c='EFG'
ejercicio_4_d='BDFH'
ejercicio_4_e='A'
ejercicio_5_a=[[1,2],[3,4]]
ejercicio_5_b=[[1,2,3],[3,5,6]]
ejercicio_5_c='no se puede sumar porque las matrices son de diferente tamaño'
ejercicio_5_d=[[1,2],[3,5],[3,5]]
ejercicio_5_e='no se ...'
ejercicio_5_f=[[1,2],[3,5]]
ejercicio_10_a=[[1,2],[3,5]]
ejercicio_10_b=[[1,2],[3,5]]
ejercicio_10_c=[[1,2],[3,5]]
ejercicio_10_d=[[1,2],[3,5]]
ejercicio_10_e=[[1,2],[3,5]]

# Ejemplo y ejercicio 2:

Sea $A=\left[ \matrix{
a+1 & 10b+1 & a+2 & 10b+2 & \\
10b+2 & a+1 & 10b+1 & a+2 & \\
a+2 & 10b+2 & a+1 & 10b+1 & \\
10b+1 & a+2 & 10b+2 & a+1 & \\
} \right]$

## 1. Reemplace $a$ y $b$ con el último y penúltimo dígito de su código, respectivamente.
Voy a responder el ejercicio con datos ficticios para ilustrar <u>la forma en que deben</u> colocar la respuesta.

In [65]:

A=sp.Matrix([
    [11, 2,  13, 3],
    [4,  15, 19, 6],
    [7,  8,  14, 19],
    [3,  -9,  5,  6],
])

A

Matrix([
[11,  2, 13,  3],
[ 4, 15, 19,  6],
[ 7,  8, 14, 19],
[ 3, -9,  5,  6]])

## 2. Escriba $M=[ A : I ]$ como una matriz de numpy . Donde $I$ es la matriz identidad de $4x4$.

In [66]:
M=sp.Matrix([
    [11, 2,  13, 3, 1, 0, 0, 0],
    [4,  15, 19, 6, 0, 1, 0, 0],
    [7,  8,  14, 19, 0, 0, 1, 0],
    [3, -9, 5, 6, 0, 0, 0, 1],
])

M

Matrix([
[11,  2, 13,  3, 1, 0, 0, 0],
[ 4, 15, 19,  6, 0, 1, 0, 0],
[ 7,  8, 14, 19, 0, 0, 1, 0],
[ 3, -9,  5,  6, 0, 0, 0, 1]])

## 3. Realice las operaciones entre renglones hasta llegar a la forma <u>escalón reducida</u> $[R:B]$. Escriba B como una matríz de numpy.

## Columna 0

In [67]:
M.row_swap(0,1)  # Se intercambia el renglón 0 por el menor de los inferiores (paso opcional)
M

Matrix([
[ 4, 15, 19,  6, 0, 1, 0, 0],
[11,  2, 13,  3, 1, 0, 0, 0],
[ 7,  8, 14, 19, 0, 0, 1, 0],
[ 3, -9,  5,  6, 0, 0, 0, 1]])

In [68]:
M[1,:] = 11*M[0,:] - 4*M[1,:]   # Se coloca un cero en el renglón 1
M[2,:] =  7*M[0,:] - 4*M[2,:]   # Se coloca un cero en el renglón 2
M[3,:] =  3*M[0,:] - 4*M[3,:]   # Se coloca un cero en el renglón 3
M

Matrix([
[4,  15,  19,   6,  0,  1,  0,  0],
[0, 157, 157,  54, -4, 11,  0,  0],
[0,  73,  77, -34,  0,  7, -4,  0],
[0,  81,  37,  -6,  0,  3,  0, -4]])

## Columna 1

In [69]:
M.row_swap(2,1)     # Se intercambia el renglón 1 por el menor de los inferiores (paso opcional)
M

Matrix([
[4,  15,  19,   6,  0,  1,  0,  0],
[0,  73,  77, -34,  0,  7, -4,  0],
[0, 157, 157,  54, -4, 11,  0,  0],
[0,  81,  37,  -6,  0,  3,  0, -4]])

In [70]:
M[2,:] = 157*M[1,:] - 73*M[2,:]   # Se coloca un cero en el renglón 2
M[3,:] =  81*M[1,:] - 73*M[3,:]   # Se coloca un cero en el renglón 3
M[0,:] =  15*M[1,:] - 73*M[0,:]   # Se coloca un cero en el renglón 0
M

Matrix([
[-292,  0, -232,  -948,   0,  32,  -60,   0],
[   0, 73,   77,   -34,   0,   7,   -4,   0],
[   0,  0,  628, -9280, 292, 296, -628,   0],
[   0,  0, 3536, -2316,   0, 348, -324, 292]])

## Columna 2

In [71]:
M[3,:] = 3536*M[2,:] - 628*M[3,:]   # Se coloca un cero en el renglón 3
M[0,:] =  232*M[2,:] + 628*M[0,:]   # Se coloca un cero en el renglón 0
M[1,:] =   77*M[2,:] - 628*M[1,:]   # Se coloca un cero en el renglón 1

M

Matrix([
[-183376,      0,   0,  -2748304,   67744,  88768,  -183376,       0],
[      0, -45844,   0,   -693208,   22484,  18396,   -45844,       0],
[      0,      0, 628,     -9280,     292,    296,     -628,       0],
[      0,      0,   0, -31359632, 1032512, 828112, -2017136, -183376]])

## Columna 3

In [72]:
M[0,:] =  2748304*M[3,:] - 31359632*M[0,:]   # Se coloca un cero en el renglón 0
M[1,:] =   693208*M[3,:] - 31359632*M[1,:]   # Se coloca un cero en el renglón 1
M[2,:] =     9280*M[3,:] - 31359632*M[2,:]   # Se coloca un cero en el renglón 2

M

Matrix([
[5750603877632,             0,            0,         0, 713229949440, -507828291328, 206900940288, -503972994304],
[            0, 1437650969408,            0,         0,  10655612608,   -2837926976,  39356157120, -127117710208],
[            0,             0, -19693848896,         0,    424698816,   -1597571712,    974826816,   -1701729280],
[            0,             0,            0, -31359632,      1032512,        828112,     -2017136,       -183376]])

## Se colocan en uno los pivotes

In [73]:
M_copia=M[:,:]

M[0,:] =  1/5750603877632*M[0,:]    # No reconoce la fracción como sp.Rational
M[1,:] =  M[1,:]*(1/1437650969408)  # No reconoce la fracción como sp.Rational
M[2,:] =  M[2,:]*1/-19693848896     # Sí reconoce la fracción como sp.Rational
M[3,:] =  M[3,:]/-31359632          # Sí reconoce la fracción como sp.Rational

M

Matrix([
[1.0,   0, 0, 0,   0.124026965622556,  -0.0883086893366606, 0.0359789936310477, -0.0876382733062684],
[  0, 1.0, 0, 0, 0.00741182166933592, -0.00197400275615479, 0.0273753212410146,  -0.088420425341726],
[  0,   0, 1, 0,          -579/26849,           2178/26849,        -1329/26849,          2320/26849],
[  0,   0, 0, 1,          -884/26849,           -709/26849,         1727/26849,           157/26849]])

In [74]:
M=M_copia # Recupera el valor anterior

M[0,:] =  M[0,:]/5750603877632   # Esta operacio es multiplicar un escalar por una matriz
M[1,:] =  M[1,:]/1437650969408
M[2,:] =  M[2,:]/-19693848896  
M[3,:] =  M[3,:]/-31359632  

M

Matrix([
[1, 0, 0, 0, 3330/26849, -2371/26849,   966/26849, -2353/26849],
[0, 1, 0, 0,  199/26849,   -53/26849,   735/26849, -2374/26849],
[0, 0, 1, 0, -579/26849,  2178/26849, -1329/26849,  2320/26849],
[0, 0, 0, 1, -884/26849,  -709/26849,  1727/26849,   157/26849]])

In [75]:
B=M[:,4:]  #De la columna 4 en adelante

B

Matrix([
[3330/26849, -2371/26849,   966/26849, -2353/26849],
[ 199/26849,   -53/26849,   735/26849, -2374/26849],
[-579/26849,  2178/26849, -1329/26849,  2320/26849],
[-884/26849,  -709/26849,  1727/26849,   157/26849]])

4. Calcule $C=AB$.


In [76]:
C=A*B
C

Matrix([
[1, 0, 0, 0],
[0, 1, 0, 0],
[0, 0, 1, 0],
[0, 0, 0, 1]])